src/core/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -354,23 +616,41 @@ function nextUnmigratedProp(doc: SavedObjectUnsanitizedDoc, migrations: ActiveMi
 function migrateProp(
   doc: SavedObjectUnsanitizedDoc,
   prop: string,
-  migrations: ActiveMigrations
-): SavedObjectUnsanitizedDoc {
+  migrations: ActiveMigrations,
+  convertTypes: boolean
+): TransformResult {
   const originalType = doc.type;
   let migrationVersion = _.clone(doc.migrationVersion) || {};
-  const typeChanged = () => !doc.hasOwnProperty(prop) || doc.type !== originalType;
+  let additionalDocs: SavedObjectUnsanitizedDoc[] = [];
+
+  for (const { version, transform, transformType } of applicableTransforms(migrations, doc, prop)) {
+    const currentVersion = propVersion(doc, prop);
+    if (currentVersion && Semver.gt(currentVersion, version)) {
+      // the previous transform function increased the object's migrationVersion; break out of the loop
+      break;
+    }
 
-  for (const { version, transform } of applicableTransforms(migrations, doc, prop)) {
-    doc = transform(doc);
-    migrationVersion = updateMigrationVersion(doc, migrationVersion, prop, version);
-    doc.migrationVersion = _.clone(migrationVersion);
+    if (transformType === 'migrate' || convertTypes) {
+      // migrate transforms are always applied, but conversion transforms and reference transforms are only applied when Kibana is upgraded
+      const result = transform(doc);
+      doc = result.transformedDoc;
+      additionalDocs = additionalDocs.concat(result.additionalDocs);
+    }
+    if (transformType === 'reference') {
+      // regardless of whether or not the reference transform was applied, increment the version
+      doc.referencesMigrationVersion = version;
```

Comment:
When we use the repository to create a saved object, it runs the document migrator on that object. If the object has a `migrations` field, any "migrate" transforms will be applied until it is up-to-date.

We don't want the same to happen WRT "convert" and "reference" transforms. However, we do want to ensure that each newly-created object's `referencesMigrationVersion` is up-to-date, so it doesn't get reference transforms applied the next time Kibana restarts.

This snippet is in place because a reference transform _exists but may not have been applied to the object_.

If we take this out we get into an endless loop because `applyMigrations` keeps calling `migrateProp` and turning this up as an applicable transform.